arrow-site/docs/docs/patterns/errorhandling/README.md

```diff
@@ -347,70 +247,74 @@ data class FormField(val label: String, val value: String)
 data class Email(val value: String)
 ```
 
-*Rules*
+*Strategies*
 
-```kotlin
-sealed class Rules<F>(A: ApplicativeError<F, Nel<ValidationError>>) : ApplicativeError<F, Nel<ValidationError>> by A {
-
-  private fun FormField.contains(needle: String): Kind<F, FormField> =
-    if (value.contains(needle, false)) just(this)
-    else raiseError(ValidationError.DoesNotContain(needle).nel())
+```kotlin:ank
+/** strategies **/
+sealed class Strategy {
+  object FailFast : Strategy()
+  object ErrorAccumulation : Strategy()
+}
 
-  private fun FormField.maxLength(maxLength: Int): Kind<F, FormField> =
-    if (value.length <= maxLength) just(this)
-    else raiseError(ValidationError.MaxLength(maxLength).nel())
+/** Abstracts away invoke strategy **/
+object Rules {
 
-  fun FormField.validateEmail(): Kind<F, Email> =
-    map(contains("@"), maxLength(250), {
-      Email(value)
-    }).handleErrorWith { raiseError(ValidationError.NotAnEmail(it).nel()) }
+  private fun FormField.contains(needle: String): ValidatedNel<ValidationError, FormField> =
+    if (value.contains(needle, false)) validNel()
+    else ValidationError.DoesNotContain(needle).invalidNel()
 
-  object ErrorAccumulationStrategy :
-    Rules<ValidatedPartialOf<Nel<ValidationError>>>(Validated.applicativeError(NonEmptyList.semigroup()))
+  private fun FormField.maxLength(maxLength: Int): ValidatedNel<ValidationError, FormField> =
+    if (value.length <= maxLength) validNel()
+    else ValidationError.MaxLength(maxLength).invalidNel()
 
-  object FailFastStrategy :
-    Rules<EitherPartialOf<Nel<ValidationError>>>(Either.applicativeError())
+  private fun FormField.validateErrorAccumulate(): ValidatedNel<ValidationError, Email> =
+    ValidatedNel.mapN(
+      Semigroup.nonEmptyList(), // accumulates errors in a non empty list
+      contains("@"),
+      maxLength(250)
+    ) { _, _ -> Email(value) }.handleErrorWith { ValidationError.NotAnEmail(it).invalidNel() }
 
-  companion object {
-    infix fun <A> failFast(f: FailFastStrategy.() -> A): A = f(FailFastStrategy)
-    infix fun <A> accumulateErrors(f: ErrorAccumulationStrategy.() -> A): A = f(ErrorAccumulationStrategy)
-  }
+  private fun FormField.validateFailFast(): Either<Nel<ValidationError>, Email> =
+    either.eager {
+      contains("@").bind() // fails fast on first error found
```

Comment:
Do we clarify anywhere that this comes without any conversion costs? So you can bind `Validated` as if it was an `Either` without any perf penalties?

Reply:
Added a comment in the code over that methods that explains it.